File: Java/Pathfinding-Algorithms/src/Pathfinding/PathNode.java
package Pathfinding;

import Utils.Pair;

public class PathNode
{
    public PathNode cameFromNode = null;

    public Pair<Integer, Integer> gridPosition;

    public int gCost = 999999;
    public int hCost = 0;
    public int fCost = 0;

    public boolean walkable = true;
    public boolean usable = true;

    public void initialize (int x, int y) {
        gridPosition = new Pair<>(x, y);
        initialize();
    }

    public void initialize (){
        gCost = 999999;
        cameFromNode = null;
        usable = true;
        CalculateFCost();
    }

    public void CalculateFCost()
    {
        fCost = gCost + hCost;
    }

    public Pair<Integer, Integer> getGridPosition() {
        return gridPosition;
    }

    public Integer getGridPosX () {
        return gridPosition.value1;
    }

    public Integer getGridPosY () {
        return gridPosition.value2;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null || getClass() != obj.getClass()) {
            return false;
        }
        PathNode other = (PathNode) obj;
        return Pair.equals(gridPosition, other.gridPosition);
    }

    public static boolean equals(PathNode a, PathNode b) {
        if (a == null && b == null) {
            return true;
        }
        if (a == null || b == null) {
            return false;
        }
        return a.equals(b);
    }

    public static boolean notEquals(PathNode a, PathNode b) {
        return !equals(a, b);
    }
}

File: C#/Pathfinding/PathNode.cs
using System.Runtime.InteropServices;

namespace Pathfinding;

public class PathNode
{
    public PathNode? cameFromNode = null; // Reference to the previous node in the search

    public Tuple<int, int> gridPosition { get; private set;  }

    public int GridPosX { get { return gridPosition.Item1; } }
    public int GridPosY { get { return gridPosition.Item2; } }
    
    public int GCost = 999999; // Walking cost from start node
    public int HCost = 0; // Heuristic cost to reach end node
    public int FCost = 0; // G cost + H cost
    
    // A public property to set if a given Node should be accessible
    public bool IsWalkable { get; set; } = true;
    public bool IsUsable { get; set; } = true;
    public bool IsChecked { get; set; } = false;

    public void Initialize(int x, int y)
    {
        gridPosition = new Tuple<int, int>(x, y);
        Initialize();
    }
    
    public void Initialize()
    {
        GCost = 999999;
        cameFromNode = null;
        CalculateFCost();
        IsUsable = true;
    }
    
    public void CalculateFCost()
    {
        FCost = GCost + HCost;
    }

    public static bool operator ==(PathNode? a, PathNode? b)
    {
        if (ReferenceEquals(a, null) && ReferenceEquals(b, null))
        {
            return true;
        }
        if (ReferenceEquals(a, null) || ReferenceEquals(b, null))
        {
            return false;
        }

        return a.gridPosition.Equals(b.gridPosition);
    }

    public static bool operator !=(PathNode? a, PathNode b)
    {
        if (ReferenceEquals(a, null) && ReferenceEquals(b, null))
        {
            return false;
        }
        if (ReferenceEquals(a, null) || ReferenceEquals(b, null))
        {
            return true;
        }

        return !a.gridPosition.Equals(b.gridPosition);
    }

}